iio: Add AccelerometerRaw support for direct access of sensor values

janbrummer opened the pull request:
Fixes: https://github.com/hadess/iio-sensor-proxy/issues/21

hadess commented on src/iio-sensor-proxy.c:
```diff
@@ -208,6 +213,10 @@ send_dbus_event (SensorData     *data,
 	}
+	if (mask & PROP_ACCELEROMETER_RAW) {
+		g_variant_builder_add (&props_builder, "{sv}", "AccelerometerRaw",
+				       g_variant_new ("(iiix)", data->previous_accel_x, data->previous_accel_y, data->previous_accel_z, g_get_real_time()));
```
That's not correct. The time needs to come from the drivers. Doing this for the evdev/input drivers is pretty straight forward, but I couldn't figure out how to do it for IIO. See #31.

janbrummer replied:
I'll have a look, but in case i cannot find a solution yet, should it be omitted?

hadess replied:
Sure, we can send out g_get_real_time() output, as we have another issue to handle this.

hadess commented on src/iio-sensor-proxy.c:
```diff
@@ -208,6 +213,10 @@ send_dbus_event (SensorData     *data,
 				       g_variant_new_string (orientation_to_string (data->previous_orientation)));
 	}
+	if (mask & PROP_ACCELEROMETER_RAW) {
+		g_variant_builder_add (&props_builder, "{sv}", "AccelerometerRaw",
```
We really don't want to be sending this as properties, as the client side would be racy. This should be sent out as a signal, so that the data is available as a single event, and we don't miss any of those events.

hadess replied:
Don't forget to also update the documentation in `net.hadess.SensorProxy.xml`!

janbrummer replied:
Ok, i'll change it, but for the record: Why is it different to e.g. LightLevel or AcclerometerOrientation?

hadess replied:
Light level and orientation don't change as often, and it's not super important if we miss a property change.

Sending the accelerometer data probably needs to happen at a much higher rate (we read every 700 msec, we probably want that position to be reported every 16msec if we want it to be usable as a controller).

hadess commented on src/iio-sensor-proxy.c:
```diff
@@ -267,6 +267,23 @@ send_dbus_event (SensorData     *data,
+
+	g_assert (data->connection);
+
+	accel = g_variant_new ("(iiix)", data->previous_accel_x, data->previous_accel_y, data->previous_accel_z, g_get_real_time());
```
A linefeed after each of the arguments, like the dbus call just below, would be nice.

hadess commented on src/iio-sensor-proxy.c:
```diff
@@ -167,11 +167,13 @@ typedef enum {
 	PROP_HAS_COMPASS                = 1 << 4,
-	PROP_COMPASS_HEADING            = 1 << 5
+	PROP_COMPASS_HEADING            = 1 << 5,
+	PROP_ACCELEROMETER_RAW          = 1 << 6,
```
Can you please squash this patch with the one following it in your branch?

hadess reviewed (commented):
We still need a way to start/stop the raw data reporting, something similar to the Claim/Release calls, but for the accelerometer raw data.

janbrummer committed:
iio: Add AccelerometerRaw support for direct access of sensor values

janbrummer commented:
I've squashed and updated the patch. It now contains Claim/ReleaseAccelerometerRaw methods and is capable of chaning poll intervall for buffered accel between NORMAL (700) and HIGH (16).

hadess commented on src/net.hadess.SensorProxy.xml:
```diff
@@ -92,6 +92,33 @@
+       accelerometer appears.
+
+       Applications should call net.hadess.SensorProxy.ReleaseAccelerometerRaw() when
+       readings are not required anymore. For example, an application that monitors
```
That's a bad example...

hadess commented on src/monitor-sensor.c:
```diff
@@ -156,14 +176,23 @@ appeared_cb (GDBusConnection *connection,
 
 	/* Accelerometer */
-	g_dbus_proxy_call_sync (iio_proxy,
+	/*g_dbus_proxy_call_sync (iio_proxy,
```
Why?

There should probably be a new option to listen to the computed and to the raw accelerometer data (maybe booleans so we can launch monitor-sensor multiple times, and check whether unwanted signals get sent to it?)

hadess commented on src/drv-iio-buffer-accel.c:
```diff
@@ -189,11 +190,33 @@ iio_buffer_accel_set_polling (gboolean state)
 }
 
+static void
+iio_buffer_accel_set_poll_interval (DriverPollInterval interval)
```
This should probably be in the core, shared between implementations.

hadess commented on src/iio-sensor-proxy.c:
```diff
@@ -258,6 +261,27 @@ send_dbus_event (SensorData     *data,
+			       g_get_real_time());
+
+	g_dbus_connection_emit_signal (data->connection,
+				       NULL,
```
We already track clients, and we'll need to start using the second paramater of `g_dbus_connection_emit_signal()` to only send this signal to the D-Bus clients that are interested in it, for efficiency reasons (no need to wake a client listening to light sensor changes with an accelerometer update), and for security reasons (the raw accelerometer information should only go to the client that requested it, not all of them).

This however poses a problem as the driver types don't allow for 2 different types of clients when raw accelerometer, or computed accelerometer data is requested).

hadess commented on src/iio-sensor-proxy.c:
```diff
@@ -362,10 +388,14 @@ handle_generic_method_call (SensorData            *data,
-			driver_set_polling (DRIVER_FOR_TYPE(driver_type), TRUE);
+		if (driver_type_exists (data, driver_type)) {
+			if (g_str_has_suffix (method_name, "Raw")) {
+				driver_set_poll_interval (DRIVER_FOR_TYPE(driver_type), DRIVER_POLL_INTERVAL_HIGH);
```
The way you've implemented this means that if a client requests a computed acceleremeter reading after a client requested raw, we'll switch the polling back. We should instead be able to see whether there are still raw clients, and if not, switch back to NORMAL.

hadess commented on src/monitor-sensor.c:
```diff
@@ -70,6 +70,23 @@ properties_changed (GDBusProxy *proxy,
+	gint x, y, z;
+	gint64 time;
+
+	if (g_strcmp0(signal_name, "AccelerometerRawUpdated"))
```
Spaces before parenthesis please. And you should add `!= 0` here. Maybe use `g_str_equal()` to make this more readable?

hadess reviewed (changes requested):
Still a lot of structural problems, but the end is in sight :)

hadess commented:
Moved to https://gitlab.freedesktop.org/hadess/iio-sensor-proxy/merge_requests/193
